File: Data_Structure/C++/SinglyCircularLL.cpp
// Singly Circular Linked List

#include<iostream>
using namespace std;
  
#pragma pack(1)
struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node* PNODE;

class SinglyCL
{
    private:
        PNODE first;
        PNODE last;
        int iCount;

    public:
      // declaration of construtor
        SinglyCL();

        void Display();
        int Count();

        void InsertFirst(int iNo);
        void InsertLast(int iNo);
        void InsertAtPos(int iNo, int iPos);

        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int iPos);

};

SinglyCL :: SinglyCL()
{
    this->first = NULL;
    this->last = NULL;

    this->iCount = 0;
}

void SinglyCL :: Display()
{
    PNODE temp = NULL;

    // Input filter
    if(first == NULL && last == NULL)
    {
        return;
    }

    temp = first;

    do
    {
        cout<<"| "<<temp->data<<" | - > ";
        temp = temp->next;

    }while(last->next != temp);

    cout<<"\n";

}

int SinglyCL :: Count()
{
    return this->iCount;
}

void SinglyCL :: InsertFirst(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL && last == NULL)
    {
        first = newn;
        last = newn;
    }
    else
    {
        newn->next = first;
        first = newn;
        
    }

    last->next = first;        // IMP
    iCount++;                  // IMP

}

void SinglyCL :: InsertLast(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL && last == NULL)
    {
        first = newn;
        last = newn;
    }
    else
    {
        last->next = newn;
        last = newn;    
    }

    last->next = first;
    iCount++;
}

void SinglyCL :: InsertAtPos(int iNo, int iPos)
{
    PNODE temp = NULL;
    PNODE newn = NULL;

    int i = 0;

    // Input Filter
    if((iPos < 1) || (iPos > iCount + 1))
    {
        cout<<"Invalid position\n";
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(iNo);
    }
    else if(iPos == iCount + 1)
    {
        InsertLast(iNo);
    }
    else
    {
        temp = first;

        newn = new NODE;

        newn->data = iNo;
        newn->next = NULL;

        for(i = 1; i < (iPos - 1); i++)
        {
            temp = temp->next;
        }

        newn->next = temp->next;
        temp->next = newn;

        iCount++;
    }

}

void SinglyCL :: DeleteFirst()   
{
    PNODE temp = NULL;

    if(first == NULL && last == NULL)
    {
        return;
    }

    else if(first == last)
    {
        delete first;

        first = NULL;
        last = NULL;
    }
    else
    {
        temp = first;

        first = first->next;

        delete temp;

        last->next = first;

    }

    iCount--;

}

void SinglyCL :: DeleteLast()
{
    PNODE temp = NULL;

    if(first == NULL && last == NULL)
    {
        return;
    }

    else if(first == last)
    {
        delete first;

        first = NULL;
        last = NULL;
    }
    else
    {
        // traversal required
       temp = first;

       while(temp->next != last)
       {
        temp = temp->next;
       }

       delete last;
       last = temp;

       last->next = first;

       iCount--;

    }

}

void SinglyCL :: DeleteAtPos(int iPos)
{
    PNODE temp = NULL;
    PNODE target = NULL;

    int i = 0;

    if((iPos < 1) || (iPos > iCount))
    {
        cout<<"Invalid Position\n";
        return;
    }

    if(iPos == 1)
    {
        DeleteFirst();
    }

    else if(iPos == iCount)
    {
        DeleteLast();
    }
    else
    {
        temp = first;

        for(i = 1; i < (iPos - 1); i++)
        {
            temp = temp->next;
        }

        target = temp->next;

        temp->next = target->next;
        delete target;
    }

    iCount--;
}


int main()
{
    SinglyCL sobj;

    int iChoice = 0;
    int iValue = 0;
    int iRet = 0;
    int iPosition = 0;

    while(iChoice != 9)                     
    {
        cout<<"---------------------------------------\n";
        cout<<"Enter your choice : \n";
        cout<<"---------------------------------------\n";

        cout<<"1 : Insert node at first position : \n";
        cout<<"2 : Insert node at last position : \n";
        cout<<"3 : Insert node at given position : \n";
        cout<<"4 : Delete node at first position : \n";
        cout<<"5 : Delete node at last position : \n";
        cout<<"6 : Delete node at given position : \n";
        cout<<"7 : Display elements: \n";
        cout<<"8 : Count elements : \n";
        cout<<"9 : Exit\n";
        cout<<"---------------------------------------\n";

        cin>>iChoice;             

        switch(iChoice)
        {
            case 1: 
                cout<<"Enter the value: \n";
                cin>>iValue;
                sobj.InsertFirst(iValue);
                break;

            case 2: 
                cout<<"Enter the value: \n";
                cin>>iValue;
                sobj.InsertLast(iValue);
                break;

            case 3: 
                cout<<"Enter the value: \n";
                cin>>iValue;
                cout<<"Enter position: \n";
                cin>>iPosition;
                sobj.InsertAtPos(iValue,iPosition);
                break;

            case 4: 
                
                sobj.DeleteFirst();
                break;

            case 5: 
                
                sobj.DeleteLast();
                break;

            case 6: 
                
                cout<<"Enter position: \n";
                cin>>iPosition;
                sobj.DeleteAtPos(iPosition);
                break;

            case 7: 
                
                cout<<"Elements of Singly Circular Linked List are : \n";
                sobj.Display();
                break;

            case 8: 
                
                iRet = sobj.Count();
                cout<<"Number of elements are : "<<iRet<<endl;
                break;

            case 9: 
                
                cout<<"Thank you for using Singly Circular Linked List Application\n";
                break;

            default:

                cout<<"Invalid choice\n";
                break;
            
        }
    }
  
    return 0;
}

File: Data_Structure/Java/SinglyCircularLL.java
//Singly Circular Linked List

import java.util.Scanner;

class Node
{
    public int data;
    public Node next;

    Node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyCL 
{
    private Node first;
    private Node last;
    private int iCount;

    public SinglyCL()
    {
        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    public void Display()
    {
        // Input filter
        if(first == null && last == null)
        {
            return;
        }

        Node temp = first;

        do 
        {
            System.out.print("| "+temp.data+" | - > ");
            temp = temp.next;

        }while(temp != last.next);

        System.out.println();
    }

    public int Count()
    {
        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        Node newn = new Node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else 
        {
            newn.next = first;
            first = newn;
        }

        last.next = first;

        iCount++;
    }

    public void InsertLast(int iNo)
    {
        Node newn = new Node(iNo);
       
        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else 
        {
            last.next = newn;
            last = last.next;
        }

        last.next = first;

        iCount++;
    }

    public void InsertAtPos(int iNo, int iPos)
    {
        Node temp = null;
        Node newn = null;

        int i = 0;

        // Input Filter
        if((iPos < 1) || (iPos > iCount + 1))
        {
            System.out.println("Invalid positon");
            return;
        }

        if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else if(iPos == iCount + 1)
        {
            InsertLast(iNo);
        }
        else 
        {
            temp = first;

            newn = new Node(iNo);

            for(i = 1; i < (iPos - 1); i++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }

    }

    public void DeleteFirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else 
        {
            first = first.next;

            last.next = first;
        }

        iCount--;
    }

    public void DeleteLast()
    {
        Node temp = null;

        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else 
        {
            temp = first;

            while(temp.next != last)
            {
                temp = temp.next;
            }

            last = temp;

            last.next = first;
        }

        iCount--;

    }

    public void DeleteAtPos(int iPos)
    {
        int i = 0;
        Node temp = null;

        // Input Filter
        if((iPos < 1) || (iPos > iCount))
        {
            System.out.println("Invalid position");
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == iCount)
        {
            DeleteLast();
        }
        else 
        {
            temp = first;
            Node target = null;

            for(i = 1; i < (iPos - 1); i++)
            {
                temp = temp.next;
            }

            // target initialized
            target = temp.next;

            temp.next = target.next;
            //target = null;

            iCount--;
        }

    }
}

public class SinglyCircularLL 
{
    public static void main(String A[])
    {
        Scanner sobj = new Scanner(System.in);

        SinglyCL obj = new SinglyCL();

        int iChoice = 0;
        int iValue = 0;
        int iPosition = 0;
        int iRet = 0;

        while(iChoice != 9)
        {
            System.out.println("--------------------------------");
            System.out.println("1 : Insert First");
            System.out.println("2 : Insert Last");
            System.out.println("3 : Insert At Position");
            System.out.println("4 : Delete First");
            System.out.println("5 : Delete Last");
            System.out.println("6 : Delete At Position");
            System.out.println("7 : Display");
            System.out.println("8 : Count");
            System.out.println("9 : Exit");
            System.out.println("--------------------------------");

            System.out.println("Enter your choice : ");
            iChoice = sobj.nextInt();

            switch (iChoice) 
            {
                case 1:
                    System.out.print("Enter Value : ");
                    iValue = sobj.nextInt();
                    obj.InsertFirst(iValue);
                    break;

                case 2:
                    System.out.print("Enter Value : ");
                    iValue = sobj.nextInt();
                    obj.InsertLast(iValue);
                    break;

                case 3:
                    System.out.print("Enter Value : ");
                    iValue = sobj.nextInt();
                    System.out.print("Enter Position : ");
                    iPosition = sobj.nextInt();

                    obj.InsertAtPos(iValue, iPosition);
                    break;

                case 4:
                    obj.DeleteFirst();
                    break;

                case 5:
                    obj.DeleteLast();
                    break;

                case 6:
                    System.out.print("Enter position : ");
                    iPosition = sobj.nextInt();

                    obj.DeleteAtPos(iPosition);
                    break;

                case 7:
                    obj.Display();
                    break;

                case 8:
                    iRet = obj.Count();

                    System.out.println("Number of nodes are in Singly Circular: "+iRet);
                    break;

                case 9:
                    System.out.println("Thank you for using Singly Circular Linked List Application");
                    break;
                   
                default:
                    System.out.println("Invalid choice");
                    break;
            }
        }

        sobj.close();
    }
    
}
